spec/parsers/pdf_parser_spec.rb

```diff
@@ -1,49 +1,71 @@
 require 'spec_helper'
 
 describe FormatParser::PDFParser do
-  let(:parsed_pdf) {
+
+  def parse_pdf (pdf_filename)
```

Comment:
Could we reuse the `let` for this? Defining methods in RSpec's blocks does not only look a bit odd, but it can also lead to other stranger issues.

Reply:
Take this simplified version of the spec tests:
```
require 'spec_helper'

describe FormatParser::PDFParser do
  let(:parsed_pdf) {
    subject.call(
      File.open(
        Pathname.new(fixtures_dir).join('PDF').join(pdf_file),
        'rb'
      )
    )
  }

  shared_examples :behave_like_pdf do |hash|
    let(:pdf_file) { hash.fetch(:file) }

    it 'acts as a pdf' do
      expect(parsed_pdf).not_to be_nil
      expect(parsed_pdf.nature).to eq(:document)
      expect(parsed_pdf.format).to eq(:pdf)
      expect(parsed_pdf.content_type).to eq('application/pdf')
    end
  end

  describe 'parses a PDF file' do
    describe 'various PDF versions' do
      include_examples :behave_like_pdf, file: 'Lorem Ipsum PDF 1.6.pdf'
      include_examples :behave_like_pdf, file: 'Simple PDF 2.0 file.pdf'
    end
  end
end
```

If you place a breakpoint in `pdf_parser.rb` right after this line:
`header = safe_read(io, 9)`

and run the spec tests in debug mode, you'll see that each time the value of the header is `%PDF-2.0`, even though we would expect a 1.6 and a 2.0.

Reply:
That explains it, yes. I prefer to not use `let` in this case, because I think it's valuable to have multiple `include_examples` testing various types of PDFs. I think it would be overkill to separate them into individual `describe` blocks, and it would also become less readable. I've seen [recommendations](https://medium.com/@tomkadwill/all-about-rspec-let-a3b642e08d39) that `let` should be used sparingly, ideally limited to small spec files and describe groups.

Reply:
I would argue that duplicating the use of shared examples is a clean solution; usually that repetition is addressed with dynamic constructs (like [this](https://github.com/WeTransfer/peek/blob/9853e4b7d512ec77445731b72a0c0db3f72ef36d/spec/peek/rules/mov_spec.rb#L61) one). That keeps things readable and also has the added benefit of having useful descriptions generated on failing tests (your example above will be very generic and it'll be harder to pinpoint what exactly went wrong).

Anyway, I don't want to expand even more on the topic; if you're not convinced, feel free to choose either path ✌️